### PR TITLE
[SALTO-1876] retry 400 errors on salesforce client

### PR DESCRIPTION
**Pull request by yanaise**
On some rare cases salesforce client fails with 400, and a retry would fix the failure, so we will retry on 400

---

_Additional context for reviewer_

---
_Release Notes_: 

---
_User Notifications_:

### CODE REVIEW

File: packages/salesforce-adapter/src/client/client.ts [hunk 1]
```diff
@@ -373,6 +379,9 @@ export default class SalesforceClient {
     this.credentials = credentials
     this.config = config
     this.retryOptions = createRetryOptions(_.defaults({}, config?.retry, DEFAULT_RETRY_OPTS))
+    if (!config?.retry?.retryStrategy) {
```
Comment (netama) [inline packages/salesforce-adapter/src/client/client.ts:382]: not sure about this part, I think this should be controllable as well (we shouldn't have a strategy that the user can't select, unless there's a good reason for it). 
maybe we should just say explicitly in the docs that we consider 400 part of the network errors, to avoid having too many cases? wdyt @ori-moisis ?

File: packages/salesforce-adapter/test/client.test.ts [hunk 2]
```diff
@@ -156,6 +155,17 @@ describe('salesforce client', () => {
   })
 
   describe('with other errors ', () => {
+    it('retries with 400 error', async () => {
```
Comment (netama) [inline packages/salesforce-adapter/test/client.test.ts:158]: I don't think this should be under `other errors`, probably either with its own `describe` or as part of `with network errors` (depending on how we control the strategy, see previous comment).
either way, we should probably also check the return value and not only the dodo scope?

**Review by netama — COMMENTED** [review]
thanks for getting this done so quickly :) generally looks right but added a couple of comments, re how it should be controlled + the unittest

File: packages/salesforce-adapter/src/client/client.ts (hunk 1 above)
Comment (yanaise) [reply packages/salesforce-adapter/src/client/client.ts:382]: The user can select this by not putting anything in the configuration. Any other currently available option is selectable by the appropriate string.

**Review by ori-moisis — COMMENTED** [review]
oh no, everything I wrote was in "pending" for a week :(

File: packages/salesforce-adapter/src/client/client.ts (hunk 1 above)
Comment (netama) [reply packages/salesforce-adapter/src/client/client.ts:382]: I'm for 2 (that's what I tried to write above too...)

File: packages/salesforce-adapter/src/client/client.ts [hunk 3]
```diff
@@ -324,6 +325,11 @@ export const loginFromCredentialsAndReturnOrgId = async (
   return identityInfo.organization_id
 }
 
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+const retryNetworkOr400 = (err: Error, response: IncomingMessage, body: any): boolean =>
+  RetryStrategies.NetworkError(err, response, body)
+    || response?.statusCode === 400
```
[the remark this answers is not in the document]
Comment (yanaise) [reply packages/salesforce-adapter/src/client/client.ts:331]: It's here because that's what they do in requestretry, which might be a holdover from pre-typescript days? Anyways I removed it.

File: packages/salesforce-adapter/src/client/client.ts [hunk 4]
```diff
@@ -284,12 +284,16 @@ const sendChunked = async <TIn, TOut>({
...
+    strategy(err, response, body) || response.statusCode === 400
+
 const createRetryOptions = (retryOptions: Required<ClientRetryConfig>): RequestRetryOptions => ({
   maxAttempts: retryOptions.maxAttempts,
-  retryStrategy: RetryStrategies[retryOptions.retryStrategy],
+  retryStrategy: retry400ErrorWrapper(RetryStrategies[retryOptions.retryStrategy]),
   delayStrategy: (err, _response, _body) => {
     log.error('failed to run SFDC call for reason: %s. Retrying in %ss.',
-      err.message, (retryOptions.retryDelay / 1000))
+      err?.message ?? '', (retryOptions.retryDelay / 1000))
```
Comment (ori-moisis) [inline packages/salesforce-adapter/src/client/client.ts:296]: why was this changed needed btw?

**Review by ori-moisis — APPROVED** [review]

File: packages/salesforce-adapter/src/client/client.ts (hunk 4 above)
Comment (yanaise) [reply packages/salesforce-adapter/src/client/client.ts:296]: My test didn't include a message, so this got undefined and caused an error.
I could have fixed the test, but in case there's a use case in the real world where we do get an undefined value here, I think handling the case rather than failing is preferred.
